File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.cpp
class Solution {
  public:
  int solve(string& s,int k) {
        int count = 0;
        int start = 0;
        int distinct = 0;
        vector<int>freq(26,0);
        int n = s.size();
        for(int i = 0;i<n;i++){
            freq[s[i]-'a']++;
            if(freq[s[i]-'a'] == 1)distinct++;
            while(distinct >k){
                freq[s[start]-'a']--;
                if(freq[s[start]-'a'] == 0)distinct--;
                start++;
            }
            count += i-start+1;
        }
        return count;
    }
    int countSubstr(string& s, int k) {
        return solve(s,k)-solve(s,k-1);
    }


};

File: Difficulty: Medium/Substrings with K Distinct/substrings-with-k-distinct.java
class Solution {
    public int countSubstr(String s, int k) {
        //  code here
        char[] S=s.toCharArray();
        
        return atMostK(S,k)-atMostK(S,k-1);
    }
    
    int atMostK(char[]s ,int k){
        if(k<0)return 0;
         int map[]=new int[26];
         int distinct=0;
         
         int i=0;
         int ans=0;
         for(int j=0;j<s.length;j++){
             int ch=s[j]-'a';
             
             map[ch]++;
             if(map[ch]==1) distinct++;
             
             while(i<=j && distinct>k){
                 int c=s[i]-'a';
                 map[c]--;
                 if(map[c]==0) distinct--;
                 i++;
             }
             
             ans+=(j-i+1);
         }
         return ans;
         
    }
}